libwnbd/wnbd_ioctl.cpp

```diff
@@ -139,6 +144,367 @@ DWORD WnbdOpenAdapterCMDeviceInstance(PDEVINST DeviceInstance)
     return Status;
 }
 
+DWORD RemoveWnbdInf(_In_ LPCTSTR InfName)
+{
+    DWORD Status = 0;
+    HINF HandleInf = INVALID_HANDLE_VALUE;
+    TCHAR InfData[MAX_INF_STRING_LENGTH];
+    UINT ErrorLine;
+    INFCONTEXT Context;
+    BOOLEAN Found = FALSE;
+
+    HandleInf = SetupOpenInfFile(InfName, NULL, INF_STYLE_WIN4, &ErrorLine);
+    if (HandleInf == INVALID_HANDLE_VALUE) {
+        Status = GetLastError();
+        LogError("SetupOpenInfFile failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+    if (!SetupFindFirstLine(HandleInf, INFSTR_SECT_VERSION, INFSTR_KEY_CATALOGFILE, &Context)) {
+        Status = GetLastError();
+        LogError("SetupFindFirstLine failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+    if (!SetupGetStringField(&Context, 1, InfData, ARRAYSIZE(InfData), NULL)) {
+        Status = GetLastError();
+        LogError("SetupGetStringField failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        goto failed;
+    }
+
+    /* Match the OEM inf file based on the catalog string wnbd.cat */
+    if (!wcscmp(InfData, L"wnbd.cat")) {
+        std::wstring SearchString(InfName);
+        if (!SetupUninstallOEMInf(
+            SearchString.substr(SearchString.find_last_of(L"\\") + 1).c_str(), SUOI_FORCEDELETE, 0)) {
+            Status = GetLastError();
+            LogError("SetupUninstallOEMInfA failed with "
+                "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        }
+    }
+
+failed:
+    if (HandleInf != INVALID_HANDLE_VALUE) {
+        SetupCloseInfFile(HandleInf);
+    }
+
+    return Status;
+}
+
+/* Cycle through all OEM inf files and try a best effort mode to remove all
+ * files which include the string wnbd.cat */
+VOID CleanDrivers()
+{
+    DWORD Status = 0;
+    TCHAR OemName[MAX_PATH];
+    HANDLE DirHandle = INVALID_HANDLE_VALUE;
+    WIN32_FIND_DATA OemFindData;
+
+    if (!GetWindowsDirectory(OemName, ARRAYSIZE(OemName))) {
+        Status = GetLastError();
+        LogError("GetWindowsDirectory failed with "
+            "error: %d. Error message: %s", Status, win32_strerror(Status).c_str());
+        return;
+    }
+    if (wcscat_s(OemName, ARRAYSIZE(OemName), L"\\INF\\OEM*.INF")) {
+        LogError("wcscat_s failed");
+        return;
+    }
+    DirHandle = FindFirstFile(OemName, &OemFindData);
+    if (DirHandle == INVALID_HANDLE_VALUE) {
+        /* No OEM files */
+        return;
+    }
+
+    do {
+        Status = RemoveWnbdInf(OemFindData.cFileName);
+        if (Status) {
+            LogError("Failed while trying to remove OEM file: %ls",
+                OemFindData.cFileName);
+        }
+    } while (FindNextFile(DirHandle, &OemFindData));
+
+    FindClose(DirHandle);
+
+    return;
```

Comment:
Why not propagate the errors?

Reply:
It's a best effort to remove the drivers from the driver catalog store.
We may end up in the situation in which the current driver needs a reboot to remove the device.
The inf file can't be removed until all the driver devices are removed, but we want all other drivers.
I don't see any clear advantage to propagate the error beside just logging what failed in this case.

Reply:
> We may end up in the situation in which the current driver needs a reboot to remove the device.

Would it still need a reboot, considering that we're removing all devices?

> I don't see any clear advantage to propagate the error beside just logging what failed in this case.

When the uninstall command returns 0, one might expect all WNBD devices and drivers to be gone.